### DISCUSSION TITLE
Dependencies

### DISCUSSION THREAD

**Discussion by mgcooper** [changliao1025/pyflowline#117] [Q&A]
What version of matplotlib and cartopy is installed into your working environment?

**Comment by mgcooper** [comment]
I know matplotlib was not required in prior versions but maybe that changed with the newer plotting utilities. 

matplotlib is required just to read the configuration file because of the classes import which import the _visual module. Specifically the import of basins.py, see screen shot. 

<img width="842" alt="image" src="https://user-images.githubusercontent.com/9771768/215234980-7b9fe784-7f03-4d57-b5f8-1b6c6a4ba867.png">

**Reply by changliao1025** [reply]
Within the read configuration function it calls the pybasin class, which imports the visualization functions. We will add a conditional import using the optional dependency.

**Comment by changliao1025** [comment]
@mgcooper 
We will use optional dependency to address this.

    from setuptools import setup

    setup(
        name='mypackage',
        version='0.1',
        ...
        extras_require={
            'visualization': ['matplotlib']
        }
    )

**Reply by mgcooper** [reply]
I think matlplotlib and Cartopy are required dependencies. If the visualization imports are removed from pybasin, they may be optional,  but as far as I can tell they are required to run the software, unless the configuration files are also considered optional?

**Reply by changliao1025** [reply]
I added some conditional import, which means if the optional matplotlib and cartopy are not installed, the python module will disable the visualization capability.

The model already output geojson so user can easily visualize them using QGIS. 

Besides, built-in visualization is only experimental and may cause crash. It is used for journals only right now.
